File: aoc/2024/aoc202024.hs
module AOC202024 where

import Data.Set(Set,delete,elems,empty,fromList,insert,
                intersection,member,singleton,size,union,unions)

import AOC

aoc = AOC {
    day="../../2020/input/24",
    aocTests=[
        AOCTest {
            testData=unlines [
                "sesenwnenenewseeswwswswwnenewsewsw",
                "neeenesenwnwwswnenewnwwsewnenwseswesw",
                "seswneswswsenwwnwse",
                "nwnwneseeswswnenewneswwnewseswneseene",
                "swweswneswnenwsewnwneneseenw",
                "eesenwseswswnenwswnwnwsewwnwsene",
                "sewnenenenesenwsewnenwwwse",
                "wenwwweseeeweswwwnwwe",
                "wsweesenenewnwwnwsenewsenwwsesesenwne",
                "neeswseenwwswnwswswnw",
                "nenwswwsewswnenenewsenwsenwnesesenew",
                "enewnwewneswsewnwswenweswnenwsenwsw",
                "sweneswneswneneenwnewenewwneswswnese",
                "swwesenesewenwneswnwwneseswwne",
                "enesenwswwswneneswsenwnewswseenwsese",
                "wnwnesenesenenwwnenwsewesewsesesew",
                "nenewswnwewswnenesenwnesewesw",
                "eneswnwswnwsenenwnwnwwseeswneewsenese",
                "neswnwewnwnwseenwseesewsenwsweewe",
                "wseweeenwnesenwwwswnew"
                ],
            testResult=Just "10",
            testResult2=Just "2208"
            }
        ],
    aocCode=ParallelCode {
        pcodeParse=const parse,
        pcodeParse2=const parse,
        pcodeTest=const result,
        pcodeTest2=result2,
        pcodeResult=const result,
        pcodeResult2=result2
        }
    }

follow :: (Int,Int) -> String -> (Int,Int)
follow xy [] = xy
follow (x,y) ('e':rest) = follow (x+1,y) rest
follow (x,y) ('s':'e':rest)
  | even y = follow (x,y+1) rest
  | otherwise = follow (x+1,y+1) rest
follow (x,y) ('s':'w':rest)
  | even y = follow (x-1,y+1) rest
  | otherwise = follow (x,y+1) rest
follow (x,y) ('w':rest) = follow (x-1,y) rest
follow (x,y) ('n':'w':rest)
  | even y = follow (x-1,y-1) rest
  | otherwise = follow (x,y-1) rest
follow (x,y) ('n':'e':rest)
  | even y = follow (x,y-1) rest
  | otherwise = follow (x+1,y-1) rest

flipTile :: (Int,Int) -> Set (Int,Int) -> Set (Int,Int)
flipTile tile floor
  | member tile floor = delete tile floor
  | otherwise = insert tile floor

parse :: String -> Set (Int,Int)
parse = foldr flipTile empty . map (follow (0,0)) . lines 

result = size

neighbors :: (Int,Int) -> [(Int,Int)]
neighbors (x,y)
  | even y = [(x+1,y),(x,y+1),(x-1,y+1),(x-1,y),(x-1,y-1),(x,y-1)]
  | otherwise = [(x+1,y),(x+1,y+1),(x,y+1),(x-1,y),(x,y-1),(x+1,y-1)]

step :: Int -> Set (Int,Int) -> Set (Int,Int)
step ncpu floor =
    parallelMapReduce ncpu isBlack unions $ elems $ union floor candidates
  where
    candidates = fromList $ concatMap neighbors $ elems floor
    isBlack xy
      | n == 2 || (n == 1 && member xy floor) = singleton xy
      | otherwise = empty
      where n = size $ intersection floor $ fromList $ neighbors xy

result2 ncpu = size . head . drop 100 . iterate (step ncpu)

File: aoc/2025/aoc202024.go
package main

func init() {
	Register(&aoc202024{
		AOC: AOC{
			Day:           24,
			InputFilename: "../2020/input/24.txt",
			Tests: []Test{
				Test{
					Input: `sesenwnenenewseeswwswswwnenewsewsw
neeenesenwnwwswnenewnwwsewnenwseswesw
seswneswswsenwwnwse
nwnwneseeswswnenewneswwnewseswneseene
swweswneswnenwsewnwneneseenw
eesenwseswswnenwswnwnwsewwnwsene
sewnenenenesenwsewnenwwwse
wenwwweseeeweswwwnwwe
wsweesenenewnwwnwsenewsenwwsesesenwne
neeswseenwwswnwswswnw
nenwswwsewswnenenewsenwsenwnesesenew
enewnwewneswsewnwswenweswnenwsenwsw
sweneswneswneneenwnewenewwneswswnese
swwesenesewenwneswnwwneseswwne
enesenwswwswneneswsenwnewswseenwsese
wnwnesenesenenwwnenwsewesewsesesew
nenewswnwewswnenesenwnesewesw
eneswnwswnwsenenwnwnwwseeswneewsenese
neswnwewnwnwseenwseesewsenwsweewe
wseweeenwnesenwwwswnew
`,
					Part1: "10",
					Part2: "2208",
				},
			},
		},
	})
}

type aoc202024 struct {
	AOC
}

func (aoc *aoc202024) parse(input *Input) map[XY]uint8 {
	tiles := map[XY]uint8{}
	ns := 0
	xy := XY{}
	for ch := range input.Chars() {
		switch ch {
		case '\n':
			if tiles[xy] == 1 {
				delete(tiles, xy)
			} else {
				tiles[xy] = 1
			}
			if ns != 0 {
				panic("bad input")
			}
			xy = XY{}
		case 'n':
			if ns != 0 {
				panic("bad input")
			}
			ns = -1
		case 's':
			if ns != 0 {
				panic("bad input")
			}
			ns = 1
		case 'e':
			switch ns {
			case -1:
				xy[0] += xy[1] & 1
				xy[1]--
				ns = 0
			case 0:
				xy[0]++
			case 1:
				xy[0] += xy[1] & 1
				xy[1]++
				ns = 0
			default:
				panic("?")
			}
		case 'w':
			switch ns {
			case -1:
				xy[0] -= 1 - xy[1]&1
				xy[1]--
				ns = 0
			case 0:
				xy[0]--
			case 1:
				xy[0] -= 1 - xy[1]&1
				xy[1]++
				ns = 0
			default:
				panic("?")
			}
		default:
			panic("bad input")
		}
	}
	return tiles
}

func (aoc *aoc202024) Part1(input *Input) string {
	return IntResult(len(aoc.parse(input)))
}

func (aoc *aoc202024) Part2(input *Input) string {
	tiles := aoc.parse(input)
	adjacent := map[XY]bool{}
	next := map[XY]uint8{}
	for range 100 {
		clear(adjacent)
		for xy := range tiles {
			dx := -1 + 2*(xy[1]&1)
			adjacent[xy] = true
			adjacent[XY{xy[0] + 1, xy[1]}] = true
			adjacent[XY{xy[0] - 1, xy[1]}] = true
			adjacent[XY{xy[0], xy[1] + 1}] = true
			adjacent[XY{xy[0], xy[1] - 1}] = true
			adjacent[XY{xy[0] + dx, xy[1] + 1}] = true
			adjacent[XY{xy[0] + dx, xy[1] - 1}] = true
		}
		clear(next)
		for xy := range adjacent {
			dx := -1 + 2*(xy[1]&1)
			n := tiles[XY{xy[0] + 1, xy[1]}] + tiles[XY{xy[0] - 1, xy[1]}] + tiles[XY{xy[0], xy[1] + 1}] + tiles[XY{xy[0], xy[1] - 1}] + tiles[XY{xy[0] + dx, xy[1] + 1}] + tiles[XY{xy[0] + dx, xy[1] - 1}]
			if n == 2 || (n == 1 && tiles[xy] == 1) {
				next[xy] = 1
			}
		}
		tiles, next = next, tiles
	}
	return IntResult(len(tiles))
}
